syntax = "proto3";

package policy_trainer;

// Service for training and inference with RL agents
service PolicyTrainerService {
  // Start training a PPO model
  rpc StartTrainingPPO(TrainingRequest) returns (TrainingResponse);
  
  // Start training an A2C model
  rpc StartTrainingA2C(TrainingRequest) returns (TrainingResponse);
  
  // Start training a DQN model
  rpc StartTrainingDQN(TrainingRequest) returns (TrainingResponse);
  
  // Get training status and progress
  rpc GetTrainingStatus(StatusRequest) returns (StatusResponse);
  
  // Perform inference with a trained model
  rpc Act(ActRequest) returns (ActResponse);
}

// Request message for starting training
message TrainingRequest {
  string experiment_id = 1;
  int32 total_timesteps = 2;
  int32 seed = 3;
  map<string, string> hyperparameters = 4;  // Optional hyperparameters
  string model_output_path = 5;
}

// Response message for training start
message TrainingResponse {
  TrainingStatus status = 1;
  string message = 2;
  string run_id = 3;
}

// Training status enum
enum TrainingStatus {
  STARTED = 0;
  FAILED = 1;
}

// Request message for status check
message StatusRequest {
  string run_id = 1;
}

// Response message for status check
message StatusResponse {
  int32 timesteps_done = 1;
  bool is_done = 2;
  string last_checkpoint_path = 3;
  string error_message = 4;
}

// Request message for action inference
message ActRequest {
  string run_id = 1;
  repeated float observation = 2;
}

// Response message for action inference
message ActResponse {
  int32 action = 1;
  bool success = 2;
  string error_message = 3;
}
